test/celeritas/optical/ScintillationGenerator.test.cc

```diff
@@ -0,0 +1,196 @@
+        params = std::make_shared<ScintillationParams>(input);
+
+        // Test step input
+        dist_.num_photons = 4;
+        dist_.step_length = 1 * units::centimeter;
+        dist_.time = 0;
+        dist_.points[StepPoint::pre].speed = units::LightSpeed(0.99);
+        dist_.points[StepPoint::post].speed = units::LightSpeed(0.99 * 0.9);
+        dist_.points[StepPoint::pre].pos = {0, 0, 0};
+        dist_.points[StepPoint::post].pos = {0, 0, 1};
+        dist_.charge = units::ElementaryCharge{-1};
+        dist_.material = OpticalMaterialId{0};
+    }
+    //! Get random number generator with clean counter
+    RandomEngine& rng()
+    {
+        rng_.reset_count();
+        return rng_;
+    }
+
+  protected:
+    // Host/device storage and reference
+    std::shared_ptr<ScintillationParams const> params;
+
+    OpticalMaterialId material{0};
+    units::ElementaryCharge charge{-1};
+
+    RandomEngine rng_;
+    OpticalDistributionData dist_;
+};
+
+//---------------------------------------------------------------------------//
+// TESTS
+//---------------------------------------------------------------------------//
+
+TEST_F(ScintillationGeneratorTest, basic)
+{
+    // Output data
+    std::vector<OpticalPrimary> storage(dist_.num_photons);
+
+    // Create the generator
+    ScintillationGenerator generate_photons(
+        dist_, params->host_ref(), make_span(storage));
+    RandomEngine& rng_engine = this->rng();
+
+    // Generate optical photons for a given input
+    auto photons = generate_photons(rng_engine);
+
+    // Check results
+    std::vector<real_type> energy;
+    std::vector<real_type> time;
+    std::vector<real_type> cos_theta;
+
+    for (auto i : range(dist_.num_photons))
+    {
+        energy.push_back(photons[i].energy.value());
+        time.push_back(photons[i].time);
+        cos_theta.push_back(
+            dot_product(photons[i].direction,
+                        dist_.points[StepPoint::post].pos
+                            - dist_.points[StepPoint::pre].pos));
+    }
+
+    const real_type expected_energy[] = {9.3561354787881e-06,
+                                         9.39574581587642e-06,
+                                         1.09240249982534e-05,
+                                         6.16620934051192e-06};
+
+    const real_type expected_time[] = {7.30250028666843e-09,
+                                       1.05142594015847e-08,
+                                       3.11699961936832e-06,
+                                       2.68409417173788e-06};
+
+    const real_type expected_cos_theta[] = {0.937735542248463,
+                                            -0.775070967887639,
+                                            0.744857640134601,
+                                            -0.748206733055997};
+
+    EXPECT_VEC_SOFT_EQ(expected_energy, energy);
+    EXPECT_VEC_SOFT_EQ(expected_time, time);
+    EXPECT_VEC_SOFT_EQ(expected_cos_theta, cos_theta);
```

Comment:
We'll need to update this for the tests to pass with real_type=float, e.g. put all this inside an `if (CELERITAS_REAL_TYPE == CELERITAS_REAL_TYPE_DOUBLE)`.

Reply:
Added.